async_service/base.py

```diff
@@ -281,17 +284,28 @@ def _common_run_task(self, task: TaskAPI) -> None:
             )
             return
 
-        if task.parent is None:
-            self.logger.debug("%s: running root task %s", self, task)
-            self._root_tasks.add(task)
-        else:
-            task.parent.add_child(task)
-            self.logger.debug("%s: %s running child task %s", self, task.parent, task)
-
+        self._add_child_task(task.parent, task)
         self._total_task_count += 1
 
         self._schedule_task(task)
 
+    def _add_child_task(self, parent, task):
+        if parent is None:
+            self.logger.debug("%s: running root task %s", self, task)
+            self._root_tasks.add(task)
+            all_children = self._root_tasks
+        else:
+            parent.add_child(task)
+            self.logger.debug("%s: %s running child task %s", self, parent, task)
+            all_children = parent.children
+
+        if len(all_children) > MAX_CHILDREN_TASKS:
```

Comment:
If you move this length check to before adding the task to the data structures above then I think you won't encounter the `AttributeError`.